include/simsycl/sycl/buffer.hh

```diff
@@ -128,34 +128,43 @@ struct buffer_access_validator {
     }
 };
 
-template<typename T, int Dimensions, typename AllocatorT>
+template<typename T, int Dimensions>
 struct buffer_state {
     using write_back_fn = std::function<void(const T *, size_t)>;
+    using deallocate_fn = std::function<void(T *, size_t)>;
+    struct raw_tag {};
 
     sycl::range<Dimensions> range;
-    AllocatorT allocator;
-    T *buffer;
-    write_back_fn write_back;
-    bool write_back_enabled;
-    std::shared_ptr<const void> shared_host_ptr; // keep the std::shared_ptr host pointer alive
+    T *data = nullptr;
+    deallocate_fn deallocate;
+    mutable shared_value<bool> write_back_on_destruction = false;
+    mutable shared_value<write_back_fn> write_back;
+    std::shared_ptr<const void> host_ptr_lifetime_extender;
     mutable shared_value<buffer_access_validator<Dimensions>> validator;
 
-    buffer_state(sycl::range<Dimensions> range, const AllocatorT &allocator = {}, const T *init_from = nullptr,
-        write_back_fn write_back = {}, const std::shared_ptr<const void> &shared_host_ptr = nullptr)
-        : range(range), allocator(allocator), buffer(this->allocator.allocate(range.size())), write_back(write_back),
-          write_back_enabled(static_cast<bool>(write_back)), shared_host_ptr(shared_host_ptr) {
+    template<typename AllocatorT>
+    buffer_state(raw_tag /* tag */, sycl::range<Dimensions> range, AllocatorT allocator, write_back_fn write_back = {},
+        std::shared_ptr<const void> lifetime_extend_host_ptr = nullptr)
+        : range(range), data(allocator.allocate(range.size())),
+          deallocate([allocator](T *ptr, size_t n) { AllocatorT(allocator).deallocate(ptr, n); }),
```

Comment:
it would be nice to have a bit of documentation here on why this whole rigamarole with the allocator is necessary.

Reply:
Added docs.